operator/crd_test.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !privileged_tests
+
+package main
+
+import (
+	"context"
+	"testing"
+
+	. "gopkg.in/check.v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	crd = &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   "cilium.io",
+			Version: "v2",
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:     "foos",
+				Singular:   "foo",
+				ShortNames: []string{"foo"},
+				Kind:       "Foo",
+			},
+			Subresources: &apiextensionsv1beta1.CustomResourceSubresources{
+				Status: &apiextensionsv1beta1.CustomResourceSubresourceStatus{},
+			},
+			Scope: apiextensionsv1beta1.ClusterScoped,
+			Validation: &apiextensionsv1beta1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{},
+			},
+		},
+	}
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) {
+	TestingT(t)
+}
+
+type crdTestSuite struct{}
+
+var _ = Suite(&crdTestSuite{})
+
+func (s *crdTestSuite) TestGetCRD(c *C) {
+	client := fake.NewSimpleClientset()
+
+	_, err := client.ApiextensionsV1beta1().CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{})
+	c.Assert(err, IsNil)
+
+	// Try to get existing CRD
+	err = WaitForCRD(client, "foo")
+	c.Assert(err, IsNil)
+
+	// Try to get non-existing CRD
+	err = waitForCRD(context.TODO(), client, "bar")
+	c.Assert(err, ErrorMatches, ".*timeout waiting for CRD bar.*")
```

Comment:
What's the timeout here? Seems like it should be the one coming from `UntilWithSync()` but I can't tell what it would be by default. Mainly just want to make sure we're not just adding arbitrary idle time into the unit test runs.

On a related note, I don't see this test name in the Travis output and when I tried to run it locally, it didn't get automatically hooked into the standard CI targets:

```
$ SKIP_VET=1 SKIP_KVSTORES=1 make unit-tests TESTPKGS=operator
dirname: missing operand
Try 'dirname --help' for more information.
  GO    tools/maptool/maptool
dirname: missing operand
Try 'dirname --help' for more information.
=== RUN   Test
OK: 1 passed
--- PASS: Test (0.00s)
=== RUN   Test_getOldestLeases
=== RUN   Test_getOldestLeases/test-1
=== RUN   Test_getOldestLeases/test-2
--- PASS: Test_getOldestLeases (0.00s)
    --- PASS: Test_getOldestLeases/test-1 (0.00s)
    --- PASS: Test_getOldestLeases/test-2 (0.00s)
=== RUN   Test_getPath
=== RUN   Test_getPath/test-1
=== RUN   Test_getPath/test-2
--- PASS: Test_getPath (0.00s)
    --- PASS: Test_getPath/test-1 (0.00s)
    --- PASS: Test_getPath/test-2 (0.00s)
PASS
coverage: 17.1% of statements in github.com/cilium/cilium/operator
ok      github.com/cilium/cilium/operator       0.037s  coverage: 17.1% of statements in github.com/cilium/cilium/operator
dirname: missing operand
Try 'dirname --help' for more information.
dirname: missing operand
Try 'dirname --help' for more information.
```

Reply:
The above assumption I made is wrong, apparently it is run, it's just that gocheck doesn't report it for whatever reason (maybe I'm missing gocheck-specific verbosity flags):

```
$ git diff
diff --git a/operator/crd_test.go b/operator/crd_test.go
index 81d4435e9a89..75b4c70229ba 100644
--- a/operator/crd_test.go
+++ b/operator/crd_test.go
@@ -73,4 +73,6 @@ func (s *crdTestSuite) TestGetCRD(c *C) {
        // Try to get non-existing CRD
        err = waitForCRD(context.TODO(), client, "bar")
        c.Assert(err, ErrorMatches, ".*timeout waiting for CRD bar.*")
+
+       c.Assert(true, Equals, false)
 }
$ SKIP_VET=1 SKIP_KVSTORES=1 make unit-tests TESTPKGS=operator
dirname: missing operand
Try 'dirname --help' for more information.
  GO    tools/maptool/maptool
dirname: missing operand
Try 'dirname --help' for more information.
=== RUN   Test

----------------------------------------------------------------------
FAIL: crd_test.go:63: crdTestSuite.TestGetCRD

crd_test.go:77:
    c.Assert(true, Equals, false)
... obtained bool = true
... expected bool = false

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (0.00s)
=== RUN   Test_getOldestLeases
=== RUN   Test_getOldestLeases/test-1
=== RUN   Test_getOldestLeases/test-2
--- PASS: Test_getOldestLeases (0.00s)
    --- PASS: Test_getOldestLeases/test-1 (0.00s)
    --- PASS: Test_getOldestLeases/test-2 (0.00s)
=== RUN   Test_getPath
=== RUN   Test_getPath/test-1
=== RUN   Test_getPath/test-2
--- PASS: Test_getPath (0.00s)
    --- PASS: Test_getPath/test-1 (0.00s)
    --- PASS: Test_getPath/test-2 (0.00s)
FAIL
coverage: 17.1% of statements in github.com/cilium/cilium/operator
FAIL    github.com/cilium/cilium/operator       0.039s
FAIL
make: *** [Makefile:225: unit-tests] Error 1
```